pkg/lsync/lrwmutex.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lsync
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// Lock states for a given resource
+const (
+	WRITELOCK = -1 + iota
+	NOLOCKS
+	READLOCKS
+)
+
+// A LRWMutex is a mutual exclusion lock with timeouts.
+type LRWMutex struct {
+	id     string
+	source string
+	state  int64
+	m      sync.Mutex // Mutex to prevent multiple simultaneous locks
+	ctx    context.Context
+}
+
+// NewLRWMutex - initializes a new lsync RW mutex.
+func NewLRWMutex(ctx context.Context) *LRWMutex {
+	return &LRWMutex{ctx: ctx}
+}
+
+// Lock holds a write lock on lm.
+//
+// If the lock is already in use, the calling go routine
+// blocks until the mutex is available.
+func (lm *LRWMutex) Lock() {
+
+	isWriteLock := true
+	lm.lockLoop(lm.id, lm.source, time.Duration(1<<63-1), isWriteLock)
+}
+
+// GetLock tries to get a write lock on lm before the timeout occurs.
+func (lm *LRWMutex) GetLock(id string, source string, timeout time.Duration) (locked bool) {
+
+	isWriteLock := true
+	return lm.lockLoop(id, source, timeout, isWriteLock)
+}
+
+// RLock holds a read lock on lm.
+//
+// If one or more read lock are already in use, it will grant another lock.
+// Otherwise the calling go routine blocks until the mutex is available.
+func (lm *LRWMutex) RLock() {
+
+	isWriteLock := false
+	lm.lockLoop(lm.id, lm.source, time.Duration(1<<63-1), isWriteLock)
+}
+
+// GetRLock tries to get a read lock on lm before the timeout occurs.
+func (lm *LRWMutex) GetRLock(id string, source string, timeout time.Duration) (locked bool) {
+
+	isWriteLock := false
+	return lm.lockLoop(id, source, timeout, isWriteLock)
+}
+
+// lockLoop will acquire either a read or a write lock
+//
+// The call will block until the lock is granted using a built-in
+// timing randomized back-off algorithm to try again until successful
+func (lm *LRWMutex) lockLoop(id, source string, timeout time.Duration, isWriteLock bool) bool {
+	start := time.Now().UTC()
+
+	retryCtx, cancel := context.WithCancel(lm.ctx)
```

Comment:
```suggestion
	retryCtx, cancel := context.WithTimeout(lm.ctx, timeout)
```
(if that is what you intended)